spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala

```diff
@@ -1029,12 +1029,20 @@ class CometSparkSessionExtensions
         var firstNativeOp = true
         newPlan.transformDown {
           case op: CometNativeExec =>
-            if (firstNativeOp) {
+            val newPlan = if (firstNativeOp) {
               firstNativeOp = false
               op.convertBlock()
             } else {
               op
             }
+
+            // If reaching leaf node, reset `firstNativeOp` to true
+            // because it will start a new block in next iteration.
+            if (op.children.isEmpty) {
+              firstNativeOp = true
+            }
+
+            newPlan
```

Comment:
Here this is the main diff that we don't have in `main` branch. Because native scan now is a real "native" operator, once we encounter it when transforming down, we need to reset `firstNativeOp` flag.